stdlib/dict.spy

```diff
@@ -0,0 +1,187 @@
+from hash import hash
+from unsafe import gc_alloc, ptr
+
+
+DKIX_EMPTY = 0
+# DKIX_DUMMY = 2 ** 32 - 1
+DKIX_DUMMY = 1 << 32 - 1
+# DKIX_ERROR = 2 ** 32 - 2
+DKIX_ERROR = 1 << 32 - 2
+# DKIX_KEY_CHANGED = 2 ** 32 - 3
+DKIX_KEY_CHANGED = 1 << 32 - 3
+
+MIN_LOG_SIZE = 6
+MAX_LOG_SIZE = 31
+MAX_FILL_RATIO = 2 / 3
+
+
+@blue.generic
+def dict(Key, Value):
+    hash_key = hash[Key]
+
+    @struct
+    class Entry:
+        # empty: bool
+        empty: i32
+        key: Key
+        value: Value
+
+
+    @struct
+    class DictData:
+        index: ptr[i32]
+        log_size: i32  # capacity
+        length: i32  # number of items stored
+        entries: ptr[Entry]
+
+
+    def new_data(log_size: i32) -> ptr[DictData]:
+        # assert MIN_LOG_SIZE <= log_size <= MAX_LOG_SIZE
+        data = gc_alloc(DictData)(1)
+        index = gc_alloc(i32)(1 << log_size)
+        data.index = index
+        data.log_size = 1 << log_size
+        data.length = 0
+        i = 0
+        while i < 1 << log_size:
+            data.index[i] = DKIX_EMPTY
+            # data.entries[i].empty = True
+            i += 1
+        return data
+
+    def capacity(self: _dict) -> i32:
+        data: ptr[DictData] = self.__ll__.data
+        return 1 << data.log_size
+
+    def mask(self: _dict) -> i32:
+        data: ptr[DictData] = self.__ll__.data
+        return 1 << data.log_size - 1
+
+    def distance_0(self: _dict, key: Key) -> i32:
+        # the first slot in the index to probe when looking up key
+        return hash_key(key) & mask(self)
+
+    def get_entry_at(self: _dict, position: i32) -> Entry:
+        data: DictData = self.__ll__.data
+        return data.entries[position]
+
+    def lookup(self: _dict, key: Key) -> i32:
+        data: ptr[DictData] = self.__ll__.data
+        d0 = distance_0(self, key)
+        distance = -1
+        while distance < capacity(self):
+            distance += 1
+            ix = data.index[(d0 + distance) & capacity(self)]
+            # [Antonio]: another confusing error:
+            #   File "/Users/dp/.pyenv/versions/3.13.2/envs/spy/lib/python3.13/site-packages/fixedint/base.py", line 391, in _f
+            #     return nt(intfunc(int(self), int(other)))
+            #               ~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
+            # ZeroDivisionError: integer modulo by zero
+            # I guess this is saying that capacity() might return 0?
```

Comment:
I think it says that someone tried to do `x % 0` (not that it *might* do modulo 0... someone actually DID that).
You get the same tb with this toy program:
```
def main() -> void:
    print(2 % 0)
```

We should detect the case and raise a proper exception/panic, see #183.

That said, I cannot really understand where it comes from, because I don't see any modulo operator in your source code, and I cannot reproduce it.
Can you got this error with a different/previous version of the code?
Or, can you show me how to reproduce?